### DISCUSSION TITLE
[charabia] Extract kvariant.rs to crate for further customization (e.g. compression)

### DISCUSSION THREAD

**Discussion by choznerol** [meilisearch/product#582] [Feedback & Feature Proposal]
Following the discussion form https://github.com/meilisearch/charabia/pull/162#pullrequestreview-1188526111

> Well, what an excellent job! Let's merge it. Then, if you want to dig deeper into creating an isolated crate for the dictionaries, I let you open a new PR for it, and I'll create a new crate name in consequence.

I've extracted the code to a simple library [`choznerol/kvariants`](https://github.com/choznerol/kvariants) with some [basic compression in build script](https://github.com/choznerol/kvariants/blob/master/build.rs). Besides publishing to crate.io, for the repo itself I'm fine with either maintaining it or transferring to [meilisearch GitHub organization](https://github.com/meilisearch). Please let me know if there is a preference!

**Comment by ManyTheFish** [comment]
Hello @choznerol, thank you a lot for your work, this will contribute to reducing the size of binaries using Charabia!
However, I have a question, the current version of your repository doesn't contain any dictionary, is it what you want?

@curquiza and @Kerollmops, this work externalizes one of the dictionaries used by Charabia in a dedicated crate that encode the dictionary into a specialized Rust structure directly usable by any rust code at building time.
This model would be reproduced for every dictionary used by Charabia in the future and we use at least 3 of them so far.
 
 what should be the best approach to integrate this?
 1) keep it decentralized letting @choznerol maintain it and reproduce this model for the other dictionaries elsewhere.
 2) put everything in the Charabia repository and create a sub-directory for each crate.
 3) create a dedicated repository on our side and create a sub-directory for each dictionary crate.
 4) something else?

**Reply by choznerol** [reply]
> However, I have a question, the current version of your repository doesn't contain any dictionary, is it what you want?

Yes, currently the [`bin/sync_dictionary`](https://github.com/choznerol/kvariants/blob/master/bin/sync_dictionaries) script need to run [before](https://github.com/choznerol/kvariants/blob/a0d0826dbbdad1365516f640189979a40e81574a/.github/workflows/ci.yml#L19-L36) `cargo build`.
I haven't found the reason that I must vendor the dictionary in the repo, and pining dictionary version is also achievable by https://github.com/choznerol/kvariants/issues/4.

**Reply by Kerollmops** [reply]
Hey @choznerol and @ManyTheFish,

Thank you for the great work. It looks pretty good, indeed!
I would prefer that we keep everything in the same repository as Charabia. It will reduce the sync time. You can, indeed, create a sub-crate by language in Charabia.

**Reply by curquiza** [reply]
Hello everyone!
I agree with @Kerollmops regarding the sync part reason
Thanks @choznerol for your involvement and your work! ❤️

**Reply by choznerol** [reply]
No problem. I will create a PR in the following days :)
